registries/helm_index.go

```
@@ -20,29 +26,54 @@ type IndexEntry struct {
+
+var clientSet kubernetes.Interface
+var settings *cli.EnvSettings
+
+func init() {
+	settings = cli.New()
+	actionConfig := new(action.Configuration)
+	if err := actionConfig.Init(settings.RESTClientGetter(), settings.Namespace(), os.Getenv("HELM_DRIVER"), log.Printf); err != nil {
+		log.WithError(err).Fatal(err)
 	}
-	defer resp.Body.Close()
 
-	index := IndexEntries{}
-	err = yaml.NewDecoder(resp.Body).Decode(&index)
+	var err error
+	clientSet, err = actionConfig.KubernetesClientSet()
 	if err != nil {
-		log.WithError(err).WithField("chart", chart).WithField("registry", r.HelmRegistry.URL).Error("Failed to unmarshal chart info")
-		return versioning.Failure
+		log.WithError(err).Fatal(err)
 	}
+}
+
+func (r HelmOverrideRegistry) getChartVersions(chart string) string {
 
-	var versions []string
-	entries := index.Entries[chart]
-	if entries == nil {
-		return versioning.Notfound
+	// trim the index.yaml suffix from the chart url, just to avoid breaking changes.
+	url := strings.TrimSuffix(r.HelmRegistry.URL, indexYamlSuffix)
+
+	entry := &repo.Entry{
+		Name: chart,
+		URL:  url,
 	}
-	for _, entry := range entries {
-		versions = append(versions, entry.Version)
+
+	if r.HelmRegistry.SecretRef != nil {
+		secrets, err := clientSet.CoreV1().Secrets(settings.Namespace()).Get(context.Background(), r.HelmRegistry.SecretRef.Name, v1.GetOptions{})
+		if err != nil {
+			log.Warning(err)
+			return versioning.Failure
+		}
+		entry.Username = string(secrets.Data[r.HelmRegistry.SecretRef.UserKey])
+		entry.Password = string(secrets.Data[r.HelmRegistry.SecretRef.PassKey])
 	}
 
-	return versioning.FindHighestVersionInList(versions, r.AllowAllReleases)
+	provider := getter.All(settings)
+
+	chartRepo, err := repo.NewChartRepository(entry, provider)
+	if err != nil {
+		log.Warning(err)
+		return versioning.Failure
+	}
+	idx, err := chartRepo.DownloadIndexFile()
+	repoIndex, err := repo.LoadIndexFile(idx)
+	chartVersion, err := repoIndex.Get(chart, "")
```

this error is being dropped.  I would think we would want to do error handling for these last 3 errors and at least log.

---

🙉 how did this slipped through! Thanks for spotting this!

---

fixed!